adapter/src/components/AuthBoundary.js

```diff
@@ -0,0 +1,112 @@
+import {
+    useConfig,
+    useDataQuery,
+    clearSensitiveCaches,
+} from '@dhis2/app-runtime'
+import { CenteredContent, NoticeBox } from '@dhis2/ui'
+import PropTypes from 'prop-types'
+import React, { useState } from 'react'
+import i18n from '../locales'
+import { LoadingMask } from './LoadingMask'
+import styles from './styles/ErrorBoundary.style.js'
+
+const LATEST_USER_KEY = 'dhis2.latestUser'
+const IS_PRODUCTION_ENV = process.env.NODE_ENV === 'production'
+
+const APP_MANAGER_AUTHORITY = 'M_dhis-web-maintenance-appmanager'
+const REQUIRED_APP_AUTHORITY = process.env.REACT_APP_DHIS2_APP_AUTH_NAME
+
+const USER_QUERY = {
+    user: {
+        resource: 'me',
+        params: { fields: ['id', 'authorities'] },
+    },
+}
+
+async function clearCachesIfUserChanged({ currentUserId, pwaEnabled }) {
+    const latestUserId = localStorage.getItem(LATEST_USER_KEY)
+    if (currentUserId !== latestUserId) {
+        const cachesCleared = await clearSensitiveCaches()
+        localStorage.setItem(LATEST_USER_KEY, currentUserId)
+        if (cachesCleared && pwaEnabled) {
+            // If this is a PWA app, the app-shell cache will need to
+            // be restored with a page reload
+            return window.location.reload()
+        }
+    }
+}
+
+const isAppAvailable = (authorities) => {
+    // Skip check on dev
+    // TODO: should we check on dev environments too?
+    if (!IS_PRODUCTION_ENV) {
+        return true
+    }
+    // Check for three possible authorities
+    return authorities.some((authority) =>
+        ['ALL', APP_MANAGER_AUTHORITY, REQUIRED_APP_AUTHORITY].includes(
+            authority
+        )
+    )
+}
+
+const ForbiddenScreen = ({ appName, baseUrl }) => {
```

Comment:
Maybe consider renaming. Perhaps something like `AccessErrorOverlay` / `ForbiddenOverlay` would be more accurate. I guess my main qualm is with the word `Screen`.